tests/src/Interop/COM/NativeServer/StringTesting.h

```diff
@@ -0,0 +1,327 @@
+        ReverseInplace(::strlen(a), a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPStr_OutAttr)(
+        /*[in,out]*/ LPSTR a,
+        /*[out]*/ LPSTR b)
+    {
+        size_t len = ::strlen(a);
+        ReverseInplace(len, a);
+        size_t byteLen = (len + 1) * sizeof(*a);
+        ::memcpy_s(b, byteLen, a, byteLen);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_LPWStr)(
+        /*[in]*/ LPWSTR a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        return Reverse(a, pRetVal);
+    }
+    DEF_RAWFUNC(Reverse_LPWStr_Ref)(
+        /*[in,out]*/ LPWSTR * a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        HRESULT hr;
+        RETURN_IF_FAILED(Reverse(*a, pRetVal));
+        ReverseInplace(::wcslen(*a), *a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_LPWStr_InRef)(
+        /*[in]*/ LPWSTR * a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        return Reverse(*a, pRetVal);
+    }
+    DEF_RAWFUNC(Reverse_LPWStr_Out)(
+        /*[in]*/ LPWSTR a,
+        /*[out]*/ LPWSTR * b)
+    {
+        return Reverse(a, b);
+    }
+    DEF_RAWFUNC(Reverse_LPWStr_OutAttr)(
+        /*[in]*/ LPWSTR a,
+        /*[out]*/ LPWSTR b)
+    {
+        if (!EqualValue(a, b))
+            return E_UNEXPECTED;
+        ReverseInplace(::wcslen(b), b);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPWStr)(
+        /*[in,out]*/ LPWSTR a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        HRESULT hr;
+        RETURN_IF_FAILED(Reverse(a, pRetVal));
+        ReverseInplace(::wcslen(a), a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPWStr_Ref)(
+        /*[in,out]*/ LPWSTR * a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        HRESULT hr;
+        RETURN_IF_FAILED(Reverse(*a, pRetVal));
+        ReverseInplace(::wcslen(*a), *a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPWStr_InRef)(
+        /*[in]*/ LPWSTR * a,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        HRESULT hr;
+        RETURN_IF_FAILED(Reverse(*a, pRetVal));
+        ReverseInplace(::wcslen(*a), *a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPWStr_Out)(
+        /*[in,out]*/ LPWSTR a,
+        /*[out]*/ LPWSTR * b)
+    {
+        HRESULT hr;
+        RETURN_IF_FAILED(Reverse(a, b));
+        ReverseInplace(::wcslen(a), a);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_SB_LPWStr_OutAttr)(
+        /*[in,out]*/ LPWSTR a,
+        /*[out]*/ LPWSTR b)
+    {
+        size_t len = ::wcslen(a);
+        ReverseInplace(len, a);
+        size_t byteLen = (len + 1) * sizeof(*a);
+        ::memcpy_s(b, byteLen, a, byteLen);
+        return S_OK;
+    }
+    DEF_RAWFUNC(Reverse_BStr)(
+        /*[in]*/ BSTR a,
+        /*[out,retval]*/ BSTR * pRetVal)
+    {
+        return ReverseBstr(a, pRetVal);
+    }
+    DEF_RAWFUNC(Reverse_BStr_Ref)(
+        /*[in,out]*/ BSTR * a,
+        /*[out,retval]*/ BSTR * pRetVal)
+    {
+        HRESULT hr;
```

Comment:
Suggestion: For ref scenario(not [in] ref), please try to valid the passing content, release its memory and create a new one